rfcs/20200929-keras-mixed-precision.md

```diff
@@ -0,0 +1,1178 @@
+* Have `name`, `compute_dtype` and `variable_dtype` be read-only
+* Raise ValueErrors instead of raising assertions
+* Give a warning if "mixed_float16" or "mixed_bfloat16" is used but not run on supported hardware
+* Allow instances of `tf.dtypes.DType` or anything convertible to a dtype with `tf.dtypes.as_dtype` to be passed to the constructor instead of a string. The `name` property will still be a string. Note for "mixed_float16" and "mixed_bfloat16", a string must be passed as there is no equivalent dtype.
+
+## The global policy
+
+There is a global policy which serves as the default policy for layers. It can be set and retrieved with:
+
+```python
+# Proposed API
+tf.keras.mixed_precision.set_global_policy(policy)
+tf.keras.mixed_precision.global_policy()
+```
+
+Currently in TensorFlow 2.3 these are exposed as:
+
+```python
+# Existing API in TensorFlow 2.3
+tf.keras.mixed_precision.experimental.set_policy(policy)
+tf.keras.mixed_precision.experimental.global_policy()
+```
+
+`set_policy()` is renamed to `set_global_policy()` to make it clear the policy is global and to be consistent with the `global_policy()` function.
+
+In addition to a policy, `set_global_policy` can take a string or any other type which can be passed to the constructor of policy. If a non-policy is passed, a new policy will be constructed and set as the global policy. E.g., the following two options are equivalent:
+
+```python
+name = ...
+
+# Option 1 (Both options are equivalent)
+policy = tf.keras.mixed_precision.Policy(name)
+tf.keras.mixed_precision.set_global_policy(policy)
+
+# Options 2
+tf.keras.mixed_precision.set_global_policy(name)
+```
+
+Option 2 will be preferred and used in tutorials. However, Option 1 is required if a user wants to use a non-default loss scale in the global policy.
+
+The global policy is the default policy for layers. When a layer is constructed, the layer’s dtype policy will be set to the global policy by default. For example:
+
+```python
+layer = tf.keras.layers.Dense()
+print(layer.dtype_policy.name)  # float32, the default dtype policy
+tf.keras.mixed_precision.set_global_policy("mixed_float16")
+layer2 = tf.keras.layers.Dense()
+print(layer2.dtype_policy.name)  # mixed_float16
+print(layer.dtype_policy.name)  # float32. A layer dtype policy never changes.
+```
+
+The global policy is only used by Keras to determine the default layer policy, and has no other purpose. The next section describes in detail how a layer uses dtype policies.
+
+To use mixed precision in a model, the global policy must be set to "mixed_float16" or "mixed_bfloat16" before the model is constructed. For many models, this is all that is required to use mixed precision.
+
+```python
+# Use mixed precision, for Volta+ GPUs
+tf.keras.mixed_precision.set_global_policy("mixed_float16")
+
+# Use mixed precision, for Cooper Lake CPUs, Ampere GPUs, or Google TPUs
+tf.keras.mixed_precision.set_global_policy("mixed_bfloat16")
+```
```

Comment:
The deprecated mixed precision graph rewrite [opted out of using mixed precision on non-optimal hardware](https://github.com/tensorflow/tensorflow/blob/28df4ade506c940f92d7e2cadca5ca4aae26c9ab/tensorflow/core/grappler/optimizers/auto_mixed_precision.cc#L2007-L2009).
I expect users will now need to handle environment detection manually or write custom code to detect the current hardware? I think this might lead to confusion when running in heterogeneous computing environments which requires users to be aware which policy is optimal on which hardware platform.

How do you feel about adding an `auto` policy (alternatively called `optimal` or `mixed`) that would autoselect between either `float32`, `mixed_float16` or `mixed_bfloat16` depending on the hardware support? This would mean that there are no code changes required when switching between CPU only, Volta GPUs or Ampere GPUs systems.

Reply:
I strongly considered this earlier. My main concern was that I believed users should be aware of whether they were using float16 or bfloat16. Users need to explicitly use a LossScaleOptimizer if using a custom training loop with float16 (when using `Model.fit`, a LossScaleOptimizer will automatically be used but otherwise the user must explicitly wrap their optimizer). Additionally, the SavedModel and checkpoint format will save the loss scale if mixed_float16 is used, and I don't think the SavedModel/checkpoint format should depend solely on the device that is used.

I added a paragraph summarizing why we don't do this.

Also, the batch size you should run at depends on the device as well, as different GPUs have different amounts of memory, so already users must be somewhat aware of their devices.

In practice, I expect most models supporting multiple devices to have a flag to choose between float32, mixed_float16, and mixed_bfloat16, and that flag is directly passed to `set_global_policy`. Within Google, we do this but also have a per-device per-model config file specifying what flags to pass, so running the GPU config file causes mixed_float16 to be passed to the flag and running the TPU config file causes mxied_bfloat16 to be passed.

Reply:
I see your point that users might need to customize other training parameters depending on the device anyway which is indeed very common.

I think the main reason why I was missing this autoselection is that I often end up having model configs which enable mixed precision by default. However, sometimes it can be useful to run a few steps locally on a CPU only machine for debugging which now would require users to remember to disable mixed precision as it would lead to greatly decreased performance on CPU (at least that was the case a few months ago when I last tested this API).
I don't think this is a huge issue since one can easily check for the existance of a GPU in user code and disable mixed precision if needed, but if this becomes a common source of confusion it might be interesting to rethink this behaviour.